packages/SwingSet/docs/vat-upgrade.md

```diff
@@ -0,0 +1,308 @@
+able to extract anything from the baggage.
+
+As v1 runs, it may need to add more data to the baggage, or remove stale
+data. The nested nature of the durable object storage graph means you can
+provide child objects to individual subsystems, to limit their authority:
+
+```js
+const fooData = makeScalarDurableMapStore();
+const barData = makeScalarDurableMapStore();
+baggage.set('foo data', fooData);
+baggage.set('bar data', barData);
+initializeFoo(fooData);
+initializeBar(barData);
+```
+
+## From Outside: the AdminNode Upgrade API
+
+The upgrade must be triggered by something holding the vat's `adminNode`
+object. This is typically the same thing that created the vat in the first
+place, but a common pattern is to hand the `adminNode` to some sort of
+governance object. The ability to upgrade a vat is also the ability to
+control its behavior, and any users who expect some particular vat behavior
+(e.g. they auditing some contract code) must take into account the
+possibility of upgrade, which means they'll care about who exactly can cause
+an upgrade to occur.
+
+Upgrades use bundlecaps, just like the initial `createVat()` call. So the
+first step of an upgrade is to install the v2 source bundle (unless it is a
+"cross-grade" that re-uses the original bundle). This uses
+`controller.validateAndInstallBundle()` from the outside, and
+`D(devices.bundle).getBundleCap()` from the inside. See docs/bundles.md for
+details.
+
+Once the governance object is holding the v2 bundlecap, it triggers the
+upgrade with `E(adminNode).upgrade(newBundlecap, options)`. This schedules
+the upgrade sequence (described above), and returns a Promise that resolves
+when the upgrade is complete. If the upgrade fails, the Promise is rejected
+and the old vat is reinstalled. An upgrade might fail because the new source
+bundle has a syntax error (preventing evaluation), or because the upgrade
+phase throws an exception or returns a Promise that rejects. It will also
+fail if the ugprade phase does not fulfill all of its obligations, such as
+leaving a durable Kind unattached.
+
+An important property of the `options` bag is `vatParameters`. This value is
+passed to the upgrade phase (how? TBD) and can be used to communicate with
+the upgrade-time code before any external messages have a chance to be
+delivered. In the Zoe ZCF vat, this is how new contract code will be
+delivered, so it can be executed (and can assume responsibility for v1
+obligations) to completion by the time the upgrade phase finishes.
+
+## From Inside: V2 Executes the Upgrade
+
+The v2 code wakes up inside the upgrade phase. The exact function signature
+is TBD, but for now let's assume that the v2 code is expected to provide the
+same `buildRootObject()` export as v1. In this case, the upgrade phase begins
+with `buildRootObject(vatPowers, vatParameters)` being called. The baggage
+might appear as `vatPowers.baggage`, and `vatParameters` will come from the
```

Comment:
I'm inclined towards having the baggage be a third parameter to `buildRootObject`.  `vatPowers` are powers granted to the vat but they tend to be generic powers, whereas the baggage belongs to the vat directly.

Reply:
aye, I'll use that as a starting point